block/src/Pos/Block/Dump.hs

```diff
@@ -0,0 +1,273 @@
+decodeBlockDump
+    :: (Bi BlockNoProof, Conduit.MonadResource m, MonadThrow m)
+    => Conduit ByteString m Block
+decodeBlockDump =
+    Lzma.decompress memlimit .| do
+        ver <- decodeDumpHeader
+        when (ver /= 0) $ throwM (UnsupportedBlockVersion ver (one 0))
+        decodeBlockStream
+  where
+    memlimit = Just (200*1024*1024)  -- 200 MB
+
+-- | Like 'decodeBlockDump', but without conduits.
+--
+-- /Throws:/ 'BlockDumpDecodeError'
+decodeBlockDump'
+    :: Bi BlockNoProof
+    => LByteString -> IO (OldestFirst [] Block)
+decodeBlockDump' dump = fmap OldestFirst $ Conduit.runResourceT $ runConduit $
+    Conduit.sourceLazy dump
+    .| decodeBlockDump
+    .| Conduit.sinkList
+
+-- | Deserialize the header of a dump and return data decoded from the
+-- header (i.e. block version).
+decodeDumpHeader
+    :: (MonadThrow m, MonadIO m)
+    => Consumer ByteString m Word16
+decodeDumpHeader =
+    Bi.awaitCbor decoder >>= \case
+        Nothing ->
+            throwM EmptyInput
+        Just (Left (Cbor.DeserialiseFailure off err)) ->
+            throwM (DumpFormatError off (toText err))
+        Just (Right x) ->
+            pure x
+  where
+    -- A decoder for everything preceding the block stream
+    decoder = do
+        Cbor.decodeListLenCanonicalOf 2        -- tuple with two elements
+        ver <- Cbor.decodeWord16Canonical      -- 1st element: version
+        Cbor.decodeListLenIndef                -- 2nd element: list
+        pure ver
+
+-- | Decode a stream of blocks.
+decodeBlockStream
+    :: forall m. (Bi BlockNoProof, MonadIO m, MonadThrow m)
+    => Conduit ByteString m Block
+decodeBlockStream = go 0
+  where
+    go :: Bi BlockNoProof => Int -> Conduit ByteString m Block
+    go i = Bi.awaitBi >>= \case
+        Nothing ->
+            pure ()
+        Just (Left (Cbor.DeserialiseFailure off err)) ->
+            throwM (BlockDecodeError i off (toText err))
+        Just (Right block) -> do
+            yield (addProof block)
+            go (i + 1)
+
+----------------------------------------------------------------------------
+-- Stripped blocks
+----------------------------------------------------------------------------
+
+-- | To make snapshots more compact, we strip proofs from blocks. This is a
+-- helper newtype for that.
+newtype GenericBlockNoProof b = GenericBlockNoProof (GenericBlock b)
+
+type GenesisBlockNoProof = GenericBlockNoProof C.GenesisBlockchain
+type MainBlockNoProof    = GenericBlockNoProof C.MainBlockchain
+
+type BlockNoProof = Either GenesisBlockNoProof MainBlockNoProof
+
+stripProof :: Block -> BlockNoProof
+stripProof (Left gb)  = Left  $ GenericBlockNoProof gb
+stripProof (Right mb) = Right $ GenericBlockNoProof mb
+
+addProof :: BlockNoProof -> Block
+addProof (Left  (GenericBlockNoProof gb)) = Left gb
+addProof (Right (GenericBlockNoProof mb)) = Right mb
+
+-- | Encode the block without including any proofs. When decoding, recreate
+-- the proofs.
+instance ( Typeable b
```

Comment:
Any ideas how to avoid duplicating code here? This is essentially copy-paste from `core/Pos/Binary/Core/Blockchain.hs`.

Reply:
No ideas, unfortunately. This instance could be autoderived if `deriveBiSimple` was a bit smarter.